drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,617 @@
+	if (id >= data->max_streams) {
+		return -EINVAL;
+	}
+
+	dma_stm32_clear_stream_irq(dev, id);
+
+	stm32_dma_enable_stream(dma, id);
+
+	return 0;
+}
+
+static int dma_stm32_stop(struct device *dev, u32_t id)
+{
+	struct dma_stm32_data *data = dev->driver_data;
+	struct dma_stm32_stream *stream = &data->streams[id];
+	const struct dma_stm32_config *config =
+				dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+
+	if (id >= data->max_streams) {
+		return -EINVAL;
+	}
+
+	LL_DMA_DisableIT_TC(dma, table_ll_stream[id]);
+#if defined(CONFIG_DMA_STM32_V1)
+	stm32_dma_disable_fifo_irq(dma, id);
+#endif
+	dma_stm32_disable_stream(dma, id);
+	dma_stm32_clear_stream_irq(dev, id);
+
+	/* Finally, flag stream as free */
+	stream->busy = false;
+
+	return 0;
+}
+
+struct k_mem_block block;
+
+static int dma_stm32_init(struct device *dev)
+{
+	struct dma_stm32_data *data = dev->driver_data;
+	const struct dma_stm32_config *config = dev->config->config_info;
+	struct device *clk =
+		device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	if (clock_control_on(clk,
+		(clock_control_subsys_t *) &config->pclken) != 0) {
+		LOG_ERR("clock op failed\n");
+		return -EIO;
+	}
+
+	config->config_irq(dev);
+
+	int size_stream =
+		sizeof(struct dma_stm32_stream) * data->max_streams;
+	data->streams = k_malloc(size_stream);
+	if (!data->streams) {
+		LOG_ERR("HEAP_MEM_POOL_SIZE is too small");
+		return -ENOMEM;
+	}
+	memset(data->streams, 0, size_stream);
+
+	for (int i = 0; i < data->max_streams; i++) {
+		data->streams[i].busy = false;
+	}
+
+	return 0;
+}
+
+static const struct dma_driver_api dma_funcs = {
+	.reload		 = dma_stm32_reload,
+	.config		 = dma_stm32_configure,
+	.start		 = dma_stm32_start,
+	.stop		 = dma_stm32_stop,
+};
+
+#define DMA_INIT(index)							\
+static void dma_stm32_config_irq_##index(struct device *dev);		\
+									\
+const struct dma_stm32_config dma_stm32_config_##index = {		\
+	.pclken = { .bus = DT_DMA_##index##_CLOCK_BUS,			\
+		    .enr = DT_DMA_##index##_CLOCK_BITS },		\
+	.config_irq = dma_stm32_config_irq_##index,			\
+	.base = DT_DMA_##index##_BASE_ADDRESS,				\
+	.support_m2m = DT_DMA_##index##_ST_MEM2MEM,			\
+};									\
+									\
+static struct dma_stm32_data dma_stm32_data_##index = {			\
+	.max_streams = 0,						\
+	.streams = NULL,						\
+};									\
+									\
+DEVICE_AND_API_INIT(dma_##index, DT_DMA_##index##_NAME, &dma_stm32_init,\
+		    &dma_stm32_data_##index, &dma_stm32_config_##index,	\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,	\
+		    &dma_funcs)
+
+#define irq_func(chan)                                                  \
+static void dma_stm32_irq_##chan(void *arg)				\
+{									\
+	dma_stm32_irq_handler(arg, chan);				\
+}
+
+#define IRQ_INIT(dma, chan)                                             \
+do {									\
+	if (!irq_is_enabled(DT_DMA_##dma##_IRQ_##chan)) {		\
+		irq_connect_dynamic(DT_DMA_##dma##_IRQ_##chan,		\
```

Comment:
Can you remind me why this is needed?

Reply:
In some series of SoCs, DMA irq lines are shared across some channels. Static declaration will raise an error when a irq line has been connected to the driver, so it must be dynamicly checked and connected.